processing/src/main/java/org/apache/druid/frame/processor/FrameProcessor.java

```diff
@@ -0,0 +1,72 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.processor;
+
+import it.unimi.dsi.fastutil.ints.IntSet;
+import org.apache.druid.frame.channel.ReadableFrameChannel;
+import org.apache.druid.frame.channel.WritableFrameChannel;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * A FrameProcessor is like an incremental version of Runnable that operates on {@link ReadableFrameChannel} and
+ * {@link WritableFrameChannel}.
+ *
+ * It is designed to enable interleaved non-blocking work on a fixed-size thread pool. Typically, this is done using
+ * an instance of {@link FrameProcessorExecutor}.
+ */
+public interface FrameProcessor<T>
+{
+  /**
+   * List of input channels. The positions of channels in this list are used to build the {@code readableInputs} set
+   * provided to {@link #runIncrementally}.
+   */
+  List<ReadableFrameChannel> inputChannels();
+
+  /**
+   * List of output channels.
+   */
+  List<WritableFrameChannel> outputChannels();
+
+  /**
+   * Runs some of the algorithm, without blocking, and either returns a value or a set of input channels
+   * to wait for. This method is called by {@link FrameProcessorExecutor#runFully} when all output channels are
+   * writable. Therefore, it is guaranteed that each output channel can accept at least one frame.
+   *
+   * This method must not read more than one frame from each readable input channel, and must not write more than one
+   * frame to each output channel.
+   *
+   * @param readableInputs channels from {@link #inputChannels()} that are either finished or ready to read.
+   *
+   * @return either a final return value or a set of input channels to wait for. Must be nonnull.
+   */
+  ReturnOrAwait<T> runIncrementally(IntSet readableInputs) throws IOException;
+
+  /**
+   * Cleans up resources used by this worker, including signalling to input and output channels that we are
+   * done reading and writing, via {@link ReadableFrameChannel#doneReading()} and
+   * {@link WritableFrameChannel#doneWriting()}.
+   *
+   * This method may be called before the worker reports completion via {@link #runIncrementally}, especially in
+   * cases of cancellation.
+   */
+  void cleanup() throws IOException;
+}
```

Comment:
`close()`?

Reply:
Hmm. Do you mean for it to also implement Closeable? If so have some mixed feelings about this one. There's a lot of times, when working on the frame processing system itself, where I wanted to see all the places that processors get cleaned up and trace the logic to ensure it happens exactly once.

Implementing Closeable makes this difficult, because the static analysis tools I use aren't able to track cleanup that happens as the result of, say, passing a FrameProcessor into something that accepts Closeable, and then being closed by that thing-that-accepts-Closeable.

Are you aware of a better way of doing this kind of analysis? Otherwise, I'm inclined to keep this as-is, with a comment to explain this reasoning.

Reply:
We are free to call a method `close()` without implementing `Closeable`. Implementing `Closeable` is really only needed when we want to do the closing generically, as in a `Closer`. Given the special naming, my guess is we don't close these objects generically?

Rather, the thought here was that we all can predict what `close()` does, so it is a handy way to communicate to readers the purpose of the function.

At least in Eclipse, if I ask to "show all reference", I get the references to only this one class -- if the class doesn't implement `Closeable`.

Again, just a minor point; the code works either way and we can always change things later, if there is a reason.

Reply:
Ah. Yes. The problems I have with `close()` aren't the method name, it's the Closeable interface and the effects that has on static analysis. In the current version of the patch, I've left it as-is, but updated the javadoc comment to capture some of this discussion:

```
  /**
   * Closes resources used by this worker.
   *
   * Exact same concept as {@link java.io.Closeable#close()}. This interface does not extend Closeable, in order to
   * make it easier to find all places where cleanup happens. (Static analysis tools can lose the thread when Closeables
   * are closed in generic ways.)
   *
   * Implementations typically call {@link ReadableFrameChannel#close()} and
   * {@link WritableFrameChannel#close()} on all input and output channels, as well as releasing any additional
   * resources that may be held.
   *
   * In cases of cancellation, this method may be called even if {@link #runIncrementally} has not yet returned a
   * result via {@link ReturnOrAwait#returnObject}.
   */
```